### PR TITLE
Move FreeBSD to cbl-mariner and remove v12

### PR DESCRIPTION
**Pull request by am11**
Fix https://github.com/dotnet/dotnet-buildtools-prereqs-docker/issues/950

### CODE REVIEW

File: src/cbl-mariner/manifest.json [hunk 1]
```diff
@@ -394,6 +394,32 @@
             }
           ]
         },
+        {
+          "platforms": [
+            {
+              "dockerfile": "src/ubuntu/18.04/cross/freebsd/13",
```
Comment (mthalman) [inline src/cbl-mariner/manifest.json]: The path should be updated to point to the cbl-mariner directory.

File: src/cbl-mariner/manifest.json [hunk 2]
```diff
@@ -394,6 +394,32 @@
...
+              "os": "linux",
+              "osVersion": "bionic",
```
Comment (mthalman) [inline src/cbl-mariner/manifest.json:402]: Should be mariner

File: src/cbl-mariner/2.0/cross/freebsd-arm64/13/Dockerfile
Comment (mthalman) [inline src/cbl-mariner/2.0/cross/freebsd-arm64/13/Dockerfile:1]: The file's being moved to mariner but still based on ubuntu.

File: src/cbl-mariner/manifest.json (hunk 1 above)
Comment (am11) [reply src/cbl-mariner/manifest.json]: Commenting on old commit :)

**Review by wfurt — APPROVED** [review]
LGTM

File: src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,5 @@
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local
+
+ARG ROOTFS_DIR=/crossrootfs/x64
+
+RUN /scripts/eng/common/cross/build-rootfs.sh freebsd13 x64
```
Comment (mthalman) [inline src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile:5]: I believe there's a reason the script isn't executed directly within the Dockerfile. I think it leads to excessive image size. @MichaelSimons - is that right?

File: src/cbl-mariner/manifest.json [hunk 4]
```diff
@@ -394,6 +394,32 @@
...
           ]
         },
+        {
+          "platforms": [
+            {
+              "dockerfile": "src/cbl-mariner/2.0/cross/freebsd/13/amd64",
+              "os": "linux",
+              "osVersion": "cbl-mariner2.0",
+              "tags": {
+                "cbl-mariner-2.0-cross-amd64-freebsd-13-$(System:TimeStamp)-$(System:DockerfileGitCommitSha)": {},
```
Comment (mthalman) [inline src/cbl-mariner/manifest.json:404]: There's still usage of the old `ubuntu-18.04...`  tags within various repos. Will you be updating those as well?

**Comment by Thefrank** [comment]
Testing this locally, I do not think that CBL Mariner has `libbsd`/`libbsd-devel` installed by default

**Comment by am11** [comment]
> Testing this locally, I do not think that CBL Mariner has `libbsd`/`libbsd-devel` installed by default

Yes, I'm also looking into it. I can fetch the RPM from fedora or build it from source. Trying out the latter one atm. Also, libbsd depends on libmd, which is also missing, but the trail ends there. 😅

File: src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile (hunk 3 above)
Comment (am11) [reply src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile:5]: Yes, I am adding the "builder" layer since it is anyway needed for libbsd.

**Comment by Thefrank** [comment]
Found one of them. Mariner actually has `libbsd-devel` on a non-default repo:

https://packages.microsoft.com/cbl-mariner/2.0/prod/extended/x86_64/Packages/l/

**Comment by am11** [comment]
> Found one of them. Mariner actually has `libbsd-devel` on a non-default repo:

Thanks! It pulled libmd.so subdep as well.

File: src/cbl-mariner/2.0/crossdeps-builder/Dockerfile [hunk 5]
```diff
@@ -17,6 +17,12 @@ RUN tdnf install -y \
         # Rootfs build dependencies
         debootstrap
 
+# dependencies which are installed from extended repo require additional setup
+RUN wget https://packages.microsoft.com/cbl-mariner/2.0/prod/extended/"${TARGETARCH/amd64/x86_64}"/config.repo -O /etc/yum.repos.d/mariner-extended.repo && \
```
Comment (am11) [inline src/cbl-mariner/2.0/crossdeps-builder/Dockerfile:21]:
```suggestion
RUN wget https://packages.microsoft.com/cbl-mariner/2.0/prod/extended/x86_64/config.repo -O /etc/yum.repos.d/mariner-extended.repo && \
```

File: src/cbl-mariner/2.0/cross/freebsd/13/arm64/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,9 @@
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local AS builder
+ARG ROOTFS_DIR=/crossrootfs/arm64
+
+RUN /scripts/eng/common/cross/build-rootfs.sh freebsd13 arm64
+
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-llvm-local
```
Comment (am11) [inline src/cbl-mariner/2.0/cross/freebsd/13/arm64/Dockerfile:6]:
```suggestion
FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local
```

File: src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local
```
Comment (am11) [inline src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile:1]:
```suggestion
FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local AS builder
```

File: src/cbl-mariner/2.0/cross/freebsd/13/arm64/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,9 @@
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local AS builder
+ARG ROOTFS_DIR=/crossrootfs/arm64
+
+RUN /scripts/eng/common/cross/build-rootfs.sh freebsd13 arm64
+
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local
```
Comment (am11) [inline src/cbl-mariner/2.0/cross/freebsd/13/arm64/Dockerfile:6]:
```suggestion
FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-llvm-local
```

File: src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,9 @@
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local AS builder
+ARG ROOTFS_DIR=/crossrootfs/x64
+
+RUN /scripts/eng/common/cross/build-rootfs.sh freebsd13 x64
+
+FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-builder-local
```
Comment (am11) [inline src/cbl-mariner/2.0/cross/freebsd/13/amd64/Dockerfile:6]:
```suggestion
FROM mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps-llvm-local
```

File: src/cbl-mariner/2.0/crossdeps-builder/Dockerfile [hunk 10]
```diff
@@ -17,6 +17,12 @@ RUN tdnf install -y \
         # Rootfs build dependencies
         debootstrap
 
+# dependencies which are installed from extended repo require additional setup
+RUN wget https://packages.microsoft.com/cbl-mariner/2.0/prod/extended/x86_64/config.repo -O /etc/yum.repos.d/mariner-extended.repo && \
+    tdnf makecache && \
+    tdnf install -y \
+        libbsd-devel
```
Comment (am11) [inline src/cbl-mariner/2.0/crossdeps-builder/Dockerfile:24]:
```suggestion
        libarchive-devel \
        libbsd-devel
```

File: src/cbl-mariner/manifest.json (hunk 4 above)
Comment (am11) [reply src/cbl-mariner/manifest.json:404]: That's the idea. 🙂

**Review by akoeplinger — APPROVED** [review]
LGTM

**Comment by am11** [comment]
So far CBL-mariner updates haven't been picked up in https://github.com/dotnet/versions/blob/main/build-info/docker/image-info.dotnet-dotnet-buildtools-prereqs-docker-main.json

**Pull request merged**